solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <solls/LanguageServer.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ReadFile.h>
+
+#include <liblangutil/SourceReferenceExtractor.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <ostream>
+
+#include <iostream>
+#include <string>
+
+using namespace std;
+using namespace std::placeholders;
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace solidity {
+
+namespace // {{{ helpers
+{
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ASTNodeLocator : public ASTConstVisitor
+{
+private:
+	int m_pos = -1;
+	ASTNode const* m_currentNode = nullptr;
+
+public:
+	explicit ASTNodeLocator(int _pos): m_pos{_pos}
+	{
+	}
+
+	ASTNode const* closestMatch() const noexcept { return m_currentNode; }
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (_node.location().start <= m_pos && m_pos <= _node.location().end)
+		{
+			m_currentNode = &_node;
+			return true;
+		}
+		return false;
+	}
+};
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ReferenceCollector: public frontend::ASTConstVisitor
+{
+private:
+	frontend::Declaration const& m_declaration;
+	std::vector<lsp::Server::DocumentHighlight> m_result;
+
+public:
+	explicit ReferenceCollector(frontend::Declaration const& _declaration):
+		m_declaration{_declaration}
+	{
+		fprintf(stderr, "finding refs: '%s', '%s'\n",
+				_declaration.name().c_str(),
+				_declaration.location().text().c_str());
+	}
+
+	std::vector<lsp::Server::DocumentHighlight> take() { return std::move(m_result); }
+
+	bool visit(Identifier const& _identifier) override
```

Comment:
Do we also want to count qualified references, i.e. through MemberAccess?
What about references outside the expression context like `contract C is A.B`?

Reply:
> Do we also want to count qualified references, i.e. through MemberAccess?

I think we should - if the user would expect it. I am pretty sure I do not catch *all* cases (for semantic highlighting and find-all-references case), I just found one missing case. This needs more testing later I think.